syntax = "proto3";

//Comando para generar clases
//py -m grpc_tools.protoc -I. --python_out=. --grpc_python_out=. .\chat.proto

service ChatAdmin {
// ENvia multiples mensjes al servidor
rpc SendMessage(MessageRequest) returns (stream SendMesageResponse);

// Recupera múltipes mensajes del servidor _a partir de una fecha_
rpc GetMesseges(stream GetMessagesRequest) returns (stream MessagesResponse);

//crea un chat
rpc CreateChat(CreateChatRequest) returns (CreateChatResponse);

//recupera los chats en los que está incluido un usuairo
rpc GetChats(GetChatsRequest) returns (stream GetChatsResponse);

rpc UpdateMensage(MessageRequest) returns (MessagesResponse);

rpc DeleteMesage(MessageRequest) returns ();
}

message GetChatsRequest {
    string idusuario = 1;
}

message GetChatsResponse {
    Chat chat = 1;
}

message CreateChatRequest {
    repeated string idusuario = 1;
}

message CreateChatResponse {
    Chat chat = 1;
}

message GetMessagesRequest {
    Date fecha = 1;
    string idchat = 2;
}

message MessagesResponse {
    Mensaje mensaje = 1;
}

message MessageRequest {
    Mensaje mensaje = 1;
}

message SendMesageResponse {
}


message Chat {
    string idchat = 1;
    repeated Usuario usuario = 2;
}

message Usuario {
    string idusuario = 1;
    string username = 2;
}


message Mensaje {
string contenido = 1;
Date fecha = 2;
string idchat =3;
string idremitente = 4;
}

message Date {
    int32 year = 1;
    int32 month = 2;
    int32 day = 3;
  }
